### DISCUSSION TITLE
[µTP] How to move forward with supporting it?

### DISCUSSION THREAD

**Discussion by tearfur** [transmission/transmission#6933] [Ideas]
### What is the issue?

I realise this is a lot to take in, so I'm open to questions. Discussions welcomed.

## What's the problem?

Right now, the µTP library we are using, libutp, is not being actively developed. It also looks like the lead developers here are reluctant to make updates to our fork.

However, libutp has problems that I'd like addressed, but I'm not sure how the lead developers want to go about this.

1. libutp does not drop packets when the receive window drops to 0, and in case the peer for some reason does not respect our receive window, the downstream API user have no way to stop the flood of data. FYI, this plays a crucial part when implementing the download speed limit.

   If the downstream user (e.g. Transmission) doesn't want to eat up the excess data, then they can choose to discard any incoming data they cannot hold (but they won't have a reliable byte stream anymore, and the disjoint chunks of data that are left will likely make no sense), or dropping the connection altogether.

   A fix was proposed at https://github.com/transmission/libutp/pull/15 but it was denied.

2. libutp for some reason drops any incoming packets where the type of the first µTP extension (bits `[8, 16)` in the header) is 3 or larger. When a packet is dropped, the ack number cannot advance, and the peer have no choice but to keep resending the dropped packet until it gives up trying and closes the connection. I think anyone can easily see how this might put Transmission at a disadvantage in swarms.

   You can observe this happening if you turn on the trace logs. You will occasionally find `Unexpected UDP packet` logged from `tr-udp.cc`. A lot of these packets will be from libtorrent, and they are almost always 26-bytes long, and start with the bytes `11 03`. More on this later.

   Granted, libutp only supports extension types 1 and 2 (btw, type 2 is ignored by libtorrent), but I don't see any reason why it cannot just silently skip over the first extension if it is not supported, like it's already doing with any subsequent extensions in the same packet.
   
   The good news is, the only extension right now that's common in the wild and fits this criteria is type 3. It is sent by libtorrent when it disconnects with its peer, and all it does is tell the peer the reason it decided to disconnect, so likely not much data is lost despite the connection stalling at this point. Still, there's no guarantee that peers in the swarm will not penalise Transmission when they see that the µTP connection cannot be closed gracefully.

   Also, if new extensions with even larger type numbers appear, and they are used during the middle of a connection, then Transmission will see many connections geting aborted during the middle of an active transfer.

3. The code is a pain in the butt to comprehend. 🤡 

   I don't imagine we will find many more people that can work with it.

## What options do we have?

AFAICT, our 2 options are:

1. Fix libutp. It looks to be the only standalone library out there that implements µTP.

   The problems I found so far look simple to fix, but the way libutp is written makes it really hard to be sure that my fixes have no unintended ripple effects.

2. Start our own implementation. This will be a huge undertaking, but the advantage is we will have full control over it. I have ideas to write it in a modular way, so that we can add/remove extensions easily.

### Which application of Transmission?

None

### Which version of Transmission?

ec5296c8dc15c0d51014d67ef93658e5569f354c

**Comment by tearfur** [comment]
CC @ckerr @mikedld

**Comment by killemov** [comment]
@tearfur Since you are in the annoying habit of turning issues into discussions (and the other way around and unnecessarily) why make THiS ONE an issue instead of a discussion?

I would take libutp and completely rework it over time as a separate project. Also if libutp is no longer being worked on sever the ties with it. So copy (not fork) the repo and mention it as the jumping off point.

**Reply by tearfur** [reply]
Well, you had a point there, I suppose this *is* closer to a discussion than an issue. So there, and let's leave it at that.

**Comment by tearfur** [comment]
If we are to start our own implementation, I'd like it to have a BSD-like API, so that we can unify the peer I/O code into one set, instead of having to maintain 2 sets of code to cater to 2 different API styles.

I'm interested to have a go at it once we start `4.2.0` development (if I am still available by that time, of course).

**Comment by Pentaphon** [comment]
> AFAICT, our 2 options are:

Which of the two options is more feasible? Starting from scratch with our own implementation sounds ideal, but will you be able to complete it?

Forking uTP also sounds reasonable but if the code is not easily understood by any contributor then our own implementation is obviously the better option for current and future contributors to take over.

**Reply by tearfur** [reply]
Well, I'm confident that I can write it. But whether I can write it in an easily understandable way is another story, I will need to have a go at it to get a better grasp.

**Reply by Pentaphon** [reply]
The only thing I would suggest is that you don't make this a high priority. I personally disable uTP myself and a lot of people who care about speed do as well. I also think Transmission has other undiscovered optimizations that have yet to be made, along with the crown jewel of https://github.com/transmission/transmission/issues/2462 . I'd wait until the 4.x.x nears end of development before rewriting uTP but of course, it's up to you. Thanks for being willing to take on such a huge task.

**Reply by tearfur** [reply]
One thing for sure is #2462 is not the cause of the speed regressions from `3.00` to `4.0.0`, because the same problem exists in `3.00`.

**Reply by Pentaphon** [reply]
> is not the cause of the speed regressions from `3.00` to `4.0.0`, because the same problem exists in `3.00`.

Yes, I'm aware. I just meant that nonblocking disk i/o is our biggest goal to make the client faster than any version before it while the regressions we currently see are caused by a number of problems between 3.00 to 4.0.0 plus what I suspect are issues that have existed even before 3.00. You and reardonia have diligently been chipping away at them recently. 

Plus I see the speed regressions in 4.0.6 even with uTP turned off.

**Comment by reardonia** [comment]
My solution ;)

https://github.com/reardonia/transmission/tree/with-utp

**Reply by tearfur** [reply]
How does these changes help?

**Reply by reardonia** [reply]
> How does these changes help?

I was (slightly) joking.  my view is that UTP support should be a compile-time option.  The library is not maintained and had at least one long-lived and scary security bug.  I compile without it.

note that WITH_UTP is broken in mainline and has been for quite some time.  My patches fix that.  Maintainers rejected my PR but we still have references to WITH_UTP throughout the codebase. Try it...
